fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -624,3 +623,19 @@ def _flush_reduce_calls(self) -> None:
                 bucket.sent = True
 
         self._consume_work_handles()
+
+    def _detect_train_change(self) -> bool:
+        # Optionally check whether the trainable parameters have changed
+        trainable_mask = list(map(_trainable, self._all_params))
+
+        # - one or more parameters trainability changed
+        trainability_changed = trainable_mask != self._reference_trainable_mask
+
+        # - the whole model is not trainable but we still have grad hooks
+        trainability_changed |= not self.training and len(self._grad_hooks) > 0
```

Comment:
does this mean that grad_hooks should be greater than 0 in eval model? Not sure I understand why this should be the case.

Reply:
it was meant to detect that the trainability changed, ie. we're in eval() mode but there are grad_hooks in place so we should refresh ? it's tied to the question above, I'm not sure of the reference behavior here

Reply:
From my offline conversation with @blefaudeux to understand this better:
* We can't detect when a module switches from train->eval unless we use the presence of hooks as an indicator. 
* We refresh trainable 1) at the beginning 2) when params changes their requires_grad property 3) train<->eval switch. 

Thanks for the explanation @blefaudeux !